### DISCUSSION TITLE
403 for Collection GET and PUT API

### DISCUSSION THREAD

**Discussion by harshwardhanp-metron** [CrowdStrike/falconpy#1208] [Q&A]
Hi all,

I am trying to put some data into a collection that I created in my custom app, but when I hit the API, I get a 403 error. I have created a new API key with all the read and write access, but I am still getting the error.

This is the API I am using for storing the information. I used the sample schema from the documentation to avoid any mistakes:

https://falcon.crowdstrike.com/documentation/page/vc1f2a56/collections#d8e53b7b
1. `/customobjects/v1/collections/{collection_name}/objects/{object_key}`

**Comment by jshcodes** [comment]
Hi @harshwardhanp-metron -

Thank you for the question!

Can you provide us an example of the code being executed and the error message received?

**Comment by harshwardhanp-metron** [comment]
Hi @jshcodes ,

I am using the following code to interact with the CrowdStrike Falcon API, but I am encountering an error. The client API key and secret I am using have all the necessary permissions assigned in Falcon. Additionally, I have tried using a binary file as mentioned in the documentation.

**Below is the code I am using, followed by the error message I am receiving**:

from falconpy import CustomStorage
# Replace with your CrowdStrike API credentials
CLIENT_ID = "CLIENT_ID"
CLIENT_SECRET = "CLIENT_SECRET"
BASE_URL = "https://api.crowdstrike.com"

# Define the collection name and object key
collection_name = "StorageData"
object_key = "tadas.txt"

# Initialize the CustomObjects service
custom_objects = CustomStorage(client_id=CLIENT_ID, client_secret=CLIENT_SECRET, base_url=BASE_URL)

# Define the data you want to store
data = {
    "name": "Tiger Nixon",
    "Id": 1
}
response = custom_objects.PutObject(
    body=data,
    collection_name=collection_name,
    object_key=object_key,
    dry_run=False  # Set to True if you want to validate without execution
)
print(response)


**Error** --:
{'status_code': 403, 'headers': {'Server': 'nginx', 'Date': 'Tue, 30 Jul 2024 06:20:56 GMT', 'Content-Type': 'application/json', 'Content-Length': '231', 'Connection': 'keep-alive', 'X-Content-Type-Options': 'nosniff', 'X-Cs-Traceid': 'ad803977-08d6-4e9c-a3dc-1d851690a601', 'X-Ratelimit-Limit': '15', 'X-Ratelimit-Remaining': '14', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'}, 'body': {'meta': {'query_time': 1.28e-07, 'powered_by': 'crowdstrike-api-gateway', 'trace_id': 'ad803977-08d6-4e9c-a3dc-1d851690a601'}, 'errors': [{'code': 403, 'message': 'access denied, authorization failed'}]}}

Could you please assist in resolving this issue?

Thank you,
Harshwardhan R Patil

**Reply by mrperfectkarak-svg** [reply]
Hello, I have the same problem now, can you share how you solved this problem?

Thanks in advance

**Reply by ryanpetit** [reply]
@mrperfectkarak-svg  I chatted with Crowdstrike Support. You also have to add your Foundry App ID as a header value when instantiating the CustomStorage class like:

falcon = CustomStorage(client_id=CLIENT_ID, client_secret=CLIENT_SECRET, base_url=BASE_URL, ext_headers={"X-CS-APP-ID": FOUNDRY_APP_ID})
